pkg/controller/core.oam.dev/v1alpha2/applicationrollout/rollout_handler.go

```diff
@@ -0,0 +1,254 @@
+	h.targetWorkloads, _, _, err = targetAssemble.GroupAssembledManifests()
+	if err != nil {
+		klog.Error("appRollout targetAppRevision failed to assemble target workload", "appRollout", klog.KRef(h.appRollout.Namespace, h.appRollout.Name))
+		return err
+	}
+
+	if len(h.sourceRevName) != 0 {
+		h.sourceAppRevision = new(v1beta1.ApplicationRevision)
+		if err := h.Get(ctx, types.NamespacedName{Namespace: h.appRollout.Namespace, Name: h.sourceRevName}, h.sourceAppRevision); err != nil && apierrors.IsNotFound(err) {
+			return err
+		}
+		// construct a assemble manifest for sourceAppRevision
+		sourceAssemble := assemble.NewAppManifests(h.sourceAppRevision).
+			WithWorkloadOption(assemble.PrepareWorkloadForRollout()).
+			WithWorkloadOption(rolloutWorkloadName())
+		h.sourceWorkloads, _, _, err = sourceAssemble.GroupAssembledManifests()
+		if err != nil {
+			klog.Error("appRollout sourceAppRevision failed to assemble workloads", "appRollout", klog.KRef(h.appRollout.Namespace, h.appRollout.Name))
+			return err
+		}
+	}
+	return nil
+}
+
+// we only support one workload now, so this func is to determine witch component is need to rollout
+func (h *rolloutHandler) determineRolloutComponent() error {
+	componentList := h.appRollout.Spec.ComponentList
+	// if user not set ComponentList in AppRollout we also find a common component between source and target
+	if len(componentList) == 0 {
+		// we need to find a default component
+		commons := appUtil.FindCommonComponentWithManifest(h.targetWorkloads, h.sourceWorkloads)
+		if len(commons) != 1 {
+			return fmt.Errorf("cannot find a default component, too many common components: %+v", commons)
+		}
+		h.needRollComponent = commons[0]
+	} else {
+		// assume that the validator webhook has already guaranteed that there is no more than one component for now
+		// and the component exists in both the target and source app
+		h.needRollComponent = componentList[0]
+	}
+	return nil
+}
+
+// fetch source and target workload
+func (h *rolloutHandler) fetchSourceAndTargetWorkload(ctx context.Context) (*unstructured.Unstructured, *unstructured.Unstructured, error) {
+	var sourceWorkload, targetWorkload *unstructured.Unstructured
+	var err error
+	if len(h.sourceRevName) == 0 {
+		klog.Info("source app fields not filled, this is a scale operation")
+	} else if sourceWorkload, err = h.extractWorkload(ctx, *h.sourceWorkloads[h.needRollComponent]); err != nil {
+		klog.Error("specified sourceRevName bug cannot fetch source workload")
+		return nil, nil, err
+	}
+	if targetWorkload, err = h.extractWorkload(ctx, *h.targetWorkloads[h.needRollComponent]); err != nil {
+		return nil, nil, err
+	}
+	return sourceWorkload, targetWorkload, nil
+}
+
+// extractWorkload use GVK and name of workload(assembled result) to fetch real workload in cluster
+func (h *rolloutHandler) extractWorkload(ctx context.Context, workload unstructured.Unstructured) (*unstructured.Unstructured, error) {
```

Comment:
Why need this? What rollout plan controller need is the GVK+name+ns of the workload, these info are already included in the assembled results. Just pass it to rollout plan controller.

Reply:
before  the logic will extract target workload, this is more safe

Reply:
This is responsible for the rollout plan controller. During batching, if the workload is lost, and you check it here, the rollout controller will run into endless loop while rollout plan controller will be stuck in an unknown state. Rollout plan controller already has its own re-try mechanism to get resources.